### PR TITLE
Add a check for custom handle_info return types

### PR DESCRIPTION
**Pull request by elfenlaid**
Follow up on the second part of the solution mentioned in https://github.com/mtrudel/thousand_island/issues/94#issuecomment-1789449398

### CODE REVIEW

File: lib/thousand_island/handler.ex [hunk 1]
```diff
@@ -454,6 +461,11 @@ defmodule ThousandIsland.Handler do
         do_socket_close(socket, reason)
       end
 
+      # Called if a callback raises an exception.
+      def terminate(_reason, []) do
```
Comment (elfenlaid) [inline lib/thousand_island/handler.ex:465]: As the server traps exists, this terminate clause transforms exceptions from this:

```elixir
Evaluation process terminated - an exception was raised:
    ** (FunctionClauseError) no function clause matching in MyHandler.terminate/2
        lib/thousand_island/handler.ex:2: MyHandler.terminate(
            {
                %RuntimeError{message: "  The callback's `state` doesn't match the expected `{socket, state}` form.\n  Please pass a `{socket, state}` tuple from `GenServer.handle_*` callbacks.\n"}, 
                [{MyHandler, :handle_info, 2, [file: ~c"lib/thousand_island/handler.ex", line: 390, error_info: %{module: Exception}]}, {:gen_server, :try_handle_info, 3, [file: ~c"gen_server.erl", line: 1077]}, {:gen_server, :handle_msg, 6, [file: ~c"gen_server.erl", line: 1165]}, {:proc_lib, :init_p_do_apply, 3, [file: ~c"proc_lib.erl", line: 241]}]
            },
            []
        )
        (stdlib 5.0.2) gen_server.erl:1125: :gen_server.try_terminate/3
        (stdlib 5.0.2) gen_server.erl:1321: :gen_server.terminate/10
        (stdlib 5.0.2) proc_lib.erl:241: :proc_lib.init_p_do_apply/3
```

To this:

```elixir
Evaluation process terminated - an exception was raised:
        ** (RuntimeError)   The callback's `state` doesn't match the expected `{socket, state}` form.
      Please pass a `{socket, state}` tuple from `GenServer.handle_*` callbacks.
    
            lib/thousand_island/handler.ex:390: MyHandler.handle_info/2
            (stdlib 5.0.2) gen_server.erl:1077: :gen_server.try_handle_info/3
            (stdlib 5.0.2) gen_server.erl:1165: :gen_server.handle_msg/6
            (stdlib 5.0.2) proc_lib.erl:241: :proc_lib.init_p_do_apply/3
```

**Comment by elfenlaid** [comment]
I'm not sure about the best test approach. I kinda cheated and used the following snippet to trigger an exception:

```elixir
defmodule MyHandler do
  use ThousandIsland.Handler

  def handle_info(:garble_state, {_socket, state}) do
    {:noreply, state}
  end
end

{:ok, handler} = MyHandler.start_link({[], []})

send(handler, :garble_state)

send(handler, {:tcp, nil, nil})
```

**Comment by mtrudel** [comment]
I took the opportunity to clean up a couple of other places where we check return types (handle_continue, terminate clauses etc). 

LMK if you're still cool with this and we can merge it!

**Comment by elfenlaid** [comment]
> I took the opportunity to clean up a couple of other places where we check return types (handle_continue, terminate clauses etc).
> 
> LMK if you're still cool with this and we can merge it!

Whoa, thanks. It looks amazing 🙌

**Comment by mtrudel** [comment]
Thanks for the PR!

**Comment by elfenlaid** [comment]
> Thanks for the PR!

Thanks for bearing with me 🙇

**Pull request merged**